src/Sylius/Bundle/ApiBundle/Validator/Constraints/PromotionCouponEligibilityValidator.php

```diff
@@ -43,29 +47,44 @@ public function validate(mixed $value, Constraint $constraint): void
             return;
         }
 
+        /** @var PromotionCouponInterface|null $promotionCoupon */
         $promotionCoupon = $this->promotionCouponRepository->findOneBy(['code' => $value->couponCode]);
 
-        if (!$promotionCoupon instanceof PromotionCouponInterface) {
-            $this->context
-                ->buildViolation($constraint->message)
-                ->atPath('couponCode')
-                ->addViolation()
-            ;
+        if ($promotionCoupon === null) {
+            $this->addViolation($constraint->invalid, 'COUPON_INVALID');
 
             return;
         }
 
-        /** @var OrderInterface $cart */
+        $expirationDate = $promotionCoupon->getExpiresAt();
+
+        if ($expirationDate !== null && $expirationDate < new \DateTime()) {
+            $this->addViolation($constraint->expired, 'COUPON_EXPIRED');
+
+            return;
+        }
+
+        /** @var OrderInterface|null $cart */
         $cart = $this->orderRepository->findCartByTokenValue($value->orderTokenValue);
 
+        if ($cart === null) {
+            throw new \InvalidArgumentException('Cart with given token supposed to exist at this point.');
```

Comment:
Good point, I think we should have a custom exception to indicate, that `$value`'s value doesn't match expectation, but not part of the validation.

```suggestion
            Assert::notNull($cart, sprintf('Cart (tokenValue %s) not found.', $value->orderTokenValue));
```